src/Xamarin.Android.Build.Tasks/Tasks/GenerateAdditionalProviderSources.cs

```diff
@@ -0,0 +1,124 @@
+using System;
+using System.IO;
+using System.Linq;
+using Java.Interop.Tools.JavaCallableWrappers;
+using Microsoft.Android.Build.Tasks;
+using Microsoft.Build.Framework;
+
+namespace Xamarin.Android.Tasks;
+
+public class GenerateAdditionalProviderSources : AndroidTask
+{
+	public override string TaskPrefix => "GPS";
+
+	[Required]
+	public ITaskItem [] AdditionalProviderSources { get; set; } = [];
+
+	[Required]
+	public string AndroidRuntime { get; set; } = "";
+
+	public string CodeGenerationTarget { get; set; } = "";
+
+	[Required]
+	public string IntermediateOutputDirectory { get; set; } = "";
+
+	public string? OutputDirectory { get; set; }
+
+	[Required]
+	public string TargetName { get; set; } = "";
+
+	AndroidRuntime androidRuntime;
+	JavaPeerStyle codeGenerationTarget;
+
+	public override bool RunTask ()
+	{
+		androidRuntime = MonoAndroidHelper.ParseAndroidRuntime (AndroidRuntime);
+		codeGenerationTarget = MonoAndroidHelper.ParseCodeGenerationTarget (CodeGenerationTarget);
+
+		// Retrieve the stored NativeCodeGenStateObject
+		var nativeCodeGenStates = BuildEngine4.GetRegisteredTaskObjectAssemblyLocal<NativeCodeGenStateCollection> (
+			MonoAndroidHelper.GetProjectBuildSpecificTaskObjectKey (GenerateJavaStubs.NativeCodeGenStateObjectRegisterTaskKey, WorkingDirectory, IntermediateOutputDirectory),
+			RegisteredTaskObjectLifetime.Build
+		);
+
+		// We only need the first architecture, since this task is architecture-agnostic
+		var templateCodeGenState = nativeCodeGenStates.States.First ().Value;
+
+		Generate (templateCodeGenState);
+
+		return !Log.HasLoggedErrors;
+	}
+
+	void Generate (NativeCodeGenStateObject codeGenState)
+	{
+		var additionalProviders = AdditionalProviderSources.Select (p => p.ItemSpec).ToList ();
```

Comment:
Maybe `AdditionalProviderSources` could be a `string[]` here, too?

Then you'd just foreach over it below and not call `.ToList()`.